File: sdk/javascript/examples/transaction_aggregate.js
#!/usr/bin/env node

//
// Reads utf8 encoded files with names "part<id>.txt".
// Creates aggregate transaction where each files is a message inside embedded transfers.
// Signs aggregate transaction using private key provided in file specified via `--private` switch.
//

import { readContents, readPrivateKey } from './examples_utils.js';
import symbolSdk from '../src/index.js';
import yargs from 'yargs';
import fs from 'fs';
import path from 'path';
import { fileURLToPath } from 'url';

(() => {
	const { SymbolFacade } = symbolSdk.facade;

	const addEmbeddedTransfers = (facade, publicKey) => {
		const textEncoder = new TextEncoder();

		// obtain recipient from publicKey, so direct all transfers to 'self'
		const recipientAddress = facade.network.publicKeyToAddress(publicKey);
		const resourcesDirectory = path.join(path.dirname(fileURLToPath(import.meta.url)), 'resources');

		const filenames = fs.readdirSync(resourcesDirectory).filter(filename => filename.startsWith('part'));
		filenames.sort();

		return filenames.map(filename => {
			const message = readContents(path.join(resourcesDirectory, filename));
			const embeddedTransaction = facade.transactionFactory.createEmbedded({
				type: 'transfer_transaction_v1',
				signerPublicKey: publicKey,
				recipientAddress,
				// note: additional 0 byte at the beginning is added for compatibility with explorer
				// and other tools that treat messages starting with 00 byte as "plain text"
				message: new Uint8Array([0, ...textEncoder.encode(message)])
			});

			console.log(`----> ${filename} length in bytes: ${embeddedTransaction.message.length}`);
			return embeddedTransaction;
		});
	};

	const args = yargs(process.argv.slice(2))
		.demandOption('private', 'path to file with private key')
		.argv;

	const facade = new SymbolFacade('testnet');
	const keyPair = readPrivateKey(args.private);

	const embeddedTransactions = addEmbeddedTransfers(facade, keyPair.publicKey);
	const merkleHash = facade.constructor.hashEmbeddedTransactions(embeddedTransactions);

	const aggregateTransaction = facade.transactionFactory.create({
		type: 'aggregate_complete_transaction_v2',
		signerPublicKey: keyPair.publicKey,
		fee: 0n,
		deadline: 1n,
		transactionsHash: merkleHash,
		transactions: embeddedTransactions
	});

	const signature = facade.signTransaction(keyPair, aggregateTransaction);
	facade.transactionFactory.constructor.attachSignature(aggregateTransaction, signature);

	console.log(`Hash: ${facade.hashTransaction(aggregateTransaction)}\n`);
	console.log(aggregateTransaction.toString());
})();

File: sdk/cpp/examples/transaction_aggregate.py
#!/usr/bin/env python

#
# Reads utf8 encoded files with names "part<id>.txt".
# Creates aggregate transaction where each files is a message inside embedded transfers.
# Signs aggregate transaction using private key provided in file specified via `--private` switch.
#

import argparse
from pathlib import Path

from examples.examples_utils import read_contents, read_private_key
from symbolchain.facade.SymbolFacade import SymbolFacade


def add_embedded_transfers(facade, public_key):
	# obtain recipient from public_key, so direct all transfers to 'self'
	embedded_transactions = []
	recipient_address = facade.network.public_key_to_address(public_key)
	resources_directory = Path(__file__).parent / 'resources'
	for filepath in sorted(resources_directory.glob('part*.txt')):
		message = read_contents(filepath)
		embedded_transaction = facade.transaction_factory.create_embedded({
			'type': 'transfer_transaction_v1',
			'signer_public_key': public_key,
			'recipient_address': recipient_address,
			# note: additional 0 byte at the beginning is added for compatibility with explorer
			# and other tools that treat messages starting with 00 byte as "plain text"
			'message': bytes(1) + message.encode('utf8')
		})

		embedded_transactions.append(embedded_transaction)

		print(f'----> {filepath.name} length in bytes: {len(embedded_transaction.message)}')

	return embedded_transactions


def main():
	parser = argparse.ArgumentParser(description='create aggregate')
	parser.add_argument('--private', help='path to file with private key', required=True)
	args = parser.parse_args()

	facade = SymbolFacade('testnet')
	key_pair = read_private_key(args.private)

	embedded_transactions = add_embedded_transfers(facade, key_pair.public_key)
	merkle_hash = facade.hash_embedded_transactions(embedded_transactions)

	aggregate_transaction = facade.transaction_factory.create({
		'type': 'aggregate_complete_transaction_v2',
		'signer_public_key': key_pair.public_key,
		'fee': 0,
		'deadline': 1,
		'transactions_hash': merkle_hash,
		'transactions': embedded_transactions
	})

	signature = facade.sign_transaction(key_pair, aggregate_transaction)
	facade.transaction_factory.attach_signature(aggregate_transaction, signature)

	print(f'Hash: {facade.hash_transaction(aggregate_transaction)}\n')
	print(aggregate_transaction)


if __name__ == '__main__':
	main()
